src/gui/src/dbDescriptors.cpp

```diff
@@ -527,6 +529,249 @@ bool DbNetDescriptor::getBBox(std::any object, odb::Rect& bbox) const
+      const auto* target = edge->target();
+
+      if (source->layer() == layer || target->layer() == layer) {
+        int sx, sy;
+        source->xy(sx, sy);
+
+        int tx, ty;
+        target->xy(tx, ty);
+        if (sx > tx) {
+          std::swap(sx, tx);
+        }
+        if (sy > ty) {
+          std::swap(sy, ty);
+        }
+        const odb::Rect path_line(sx, sy, tx, ty);
+        if (path_line.intersects(node_pt)) {
+          connections.insert(source);
+          connections.insert(target);
+        }
+      }
+    }
+
+    node_map[node].insert(connections.begin(), connections.end());
+    for (const auto* cnode : connections) {
+      if (cnode == node) {
+        // don't insert the source node
+        continue;
+      }
+      node_map[cnode].insert(node);
+    }
+  }
+}
+
+void DbNetDescriptor::findPath(NodeMap& graph,
+                               const Node* source,
+                               const Node* sink,
+                               std::vector<odb::Point>& path) const
+{
+  // find path from source to sink using A*
+  // https://en.wikipedia.org/w/index.php?title=A*_search_algorithm&oldid=1050302256
+
+  auto distance = [](const Node* node0, const Node* node1) -> int {
+    int x0, y0;
+    node0->xy(x0, y0);
+    int x1, y1;
+    node1->xy(x1, y1);
+    return std::abs(x0 - x1) + std::abs(y0 - y1);
+  };
+
+  std::map<const Node*, const Node*> came_from;
+  std::map<const Node*, int> g_score;
+  std::map<const Node*, int> f_score;
+
+  struct DistNode {
+    const Node* node;
+    int dist;
+
+    public:
+      // used for priority queue
+      bool operator<(const DistNode& other) const { return dist > other.dist; }
+  };
+  std::priority_queue<DistNode> open_set;
+  const int source_sink_dist = distance(source, sink);
+  open_set.push({source, source_sink_dist});
+
+  for (const auto& [node, nodes] : graph) {
+    g_score[node] = std::numeric_limits<int>::max();
+    f_score[node] = std::numeric_limits<int>::max();
+  }
+  g_score[source] = 0;
+  f_score[source] = source_sink_dist;
+
+  while (!open_set.empty()) {
+    auto current = open_set.top().node;
+    open_set.pop();
+    if (current == sink) {
+      // build path
+      int x, y;
+      while (current != source) {
+        current->xy(x, y);
+        path.push_back(odb::Point(x, y));
+        current = came_from[current];
+      }
+      current->xy(x, y);
+      path.push_back(odb::Point(x, y));
+      return;
+    }
+
+    const int current_g_score = g_score[current];
+    for (const auto& neighbor : graph[current]) {
+      const int possible_g_score = current_g_score + distance(current, neighbor);
+      if (possible_g_score < g_score[neighbor]) {
+        const int new_f_score = possible_g_score + distance(neighbor, sink);
+        came_from[neighbor] = current;
+        g_score[neighbor] = possible_g_score;
+        f_score[neighbor] = new_f_score;
+
+        open_set.push({neighbor, new_f_score});
```

Comment:
This should be
 if neighbor not in openSet
                    openSet.add(neighbor)

this will fail if there are routing loops.  I guess that isn't likely.

Reply:
added a set to keep track of the nodes in the priority_queue to check for that condition.